### PR TITLE
Add support for cwl autocompletion

### PR DESCRIPTION
**Pull request by btstream**
Autocompletion using cwl files. With unclosed environment support, just like `LaTeXing` on sublime-text.

Does anyone know what licences are used for these `cwl` files? Can them be included in this package directly?

### CODE REVIEW

**Comment by msiniscalchi** [comment]
Great! I'll look into this. So long as it can be turned off :)

File: lib/latextools.coffee [hunk 1]
```diff
@@ -283,6 +297,14 @@ module.exports = Latextools =
 
         # add more here?
 
+    # if using cwl-completion, load autocomplete provider
+    if atom.config.get("latextools.commandCompletion") != 'never'
+      @requireIfNeeded ['cwl-provider']
+      @cwlProvider.loadCompletions()
```
Comment (ig0774) [inline lib/latextools.coffee]: Shouldn't this be done in `requireIfNeeded`?

File: lib/latextools.coffee [hunk 2]
```diff
@@ -283,6 +297,14 @@ module.exports = Latextools =
 
         # add more here?
 
+    # if using cwl-completion, load autocomplete provider
+    if atom.config.get("latextools.commandCompletion") != 'never'
```
Comment (ig0774) [inline lib/latextools.coffee]: Small thing, but 300--302 should probably go in `getProvider()` that way, if someone has disabled the autocompletions plugin altogether, the cwl code can be completely ignored.

**Comment by ig0774** [comment]
I really like this! The cwl completions is one of my favourite features in Sublime-world.

**Comment by ig0774** [comment]
On the environment feature bit of things:

1. Is there any way to get the completion to auto-adjust the way it does just typing `\end` like this?
test

1. `\begin{document}` `\end{document}` completion doesn't work because `\begin{document}` doesn't setup a `environment` scope. This is fine, except it means that `\end{document}` will never appear on the completion list.

1. (I realise this is probably an altogether different feature) is it possible to enable something similar to SublimeText/LaTeXTools#621, i.e., auto-completing environment names?

**Comment by btstream** [comment]
@ig0774 

1. It has been solved by triggering `auto-indent` command after insert suggestions.

2. For now, I could not find a method to solve this except doing a full text search, but it's so inefficiency.

3. I think this feature should belong to something like `fill all`

**Comment by msiniscalchi** [comment]
On the license: I think cwl files are under the GPL v2. This may be a problem. CWL files come from the Kile text editor; here's the source tree:

https://quickgit.kde.org/?p=kile.git&a=tree

The cwl files themselves are here:

https://quickgit.kde.org/?p=kile.git&a=tree&h=76867d2408c53115ed7a892f99b6895c42af8fda&hb=7b238c42580abc936816d4ea0df61d0cbbefecc4&f=src%2Fdata%2Fcomplete%2Ftex

The whole Kile project is under the GPLv2, so I assume the cwl files are, too.

I would be fine with relicensing LaTeXTools under the GPL (v2 or v3), but I don't know if that's OK with the Atom powers that be.

**Comment by ig0774** [comment]
@btstream: Thanks for addressing that.

Re 2: I realise *properly* closing the document is probably outside of the question without somehow searching the whole file. I'd be happier, though, if the regex at line 34 of ltcwl-completion.coffee didn't filter out the `\end{document}` command. Changing it to something like `^end\{(?!document)[^\}]*\}` will at least ensure that `\end{document}` appears in the auto-complete list.

Additionally, after playing around with this a bit (and also the project file stuff) I'm thinking that using `requireIfNeeded` from `getProvider` is actually a bad idea. It would be easier to simply replace `getProvider` with an implementation like this:

```coffee
getProvider: ->
  # if using cwl-completion, load autocomplete provider
  if atom.config.get("latextools.commandCompletion") != 'never'
    @cwlProvider = require './ltcwl-completion'
    @cwlProvider.loadCompletions()
    @cwlProvider
```

And to remove the `cwl-provider` block from `requireIfNeeded`. This ensures that autocomplete-plus doesn't end up unexpectedly loading a fair bit of latextools unnecessarily.

**Comment by msiniscalchi** [comment]
Just an update @btstream and @ig0774; I'm trying to contact the Kile devs, as one of the Atom devs pointed out that Kile might be dual-licensed (GPLv2 and BSD). If that's the case, we can go ahead and pull this patch. Which, btw, is growing on me :) 

If that fails, would you be OK with relicensing under the GPLv2? @ig0774 this would have to cover your other contributions to atom-latextools.

**Comment by ig0774** [comment]
I'm perfectly fine with re-licensing under the GPL, if that's what needs to happen.

**Comment by msiniscalchi** [comment]
@ig0774 and @btstream just an update: I am doing my very best to get in touch with the Kile devs, but that's proving to be, er, challenging. I'm trying the kile-devel mailing list as a last resort. If that fails, we'll have to either switch latextools to GPLv2, or create a GPLv2 package that contains the files only. [It's not clear that even just depending on that would be a problem though...]

**Comment by Evpok** [comment]
Asking myself the exact same question for latex-autocomplete, I found http://programmers.stackexchange.com/a/218348/36889

>   If a programming language interpreter has a license that is incompatible with the GPL, can I run GPL-covered programs on it?
> 
>>  When the interpreter just interprets a language, the answer is yes. The interpreted program, to the interpreter, is just data; the GPL doesn't restrict what tools you process the program with.
> 
> It would seem that the "it's just data" argument applies in your case as well. The GPL's copyleft provisions only apply when the GPL'd work and some other work form a derivative work, which doesn't appear to happen in your case.
> 
> Of course, any GPL work you distribute must stay under the GPL, but the GPL provisions will not apply to your project as a whole, because it doesn't combine with your existing work into a new derived work.

Which is of course not lawyer advice, but makes sense, although I have a hard time exegesising the linked GPL FAQ

> However, when the interpreter is extended to provide “bindings” to other facilities (often, but not necessarily, libraries), the interpreted program is effectively linked to the facilities it uses through these bindings. The JNI or Java Native Interface is an example of such a facility; libraries that are accessed in this way are linked dynamically with the Java programs that call them.

I saw [your question on kile-devel](https://sourceforge.net/p/kile/mailman/kile-devel/thread/CAPmVbyOB8BZjhut7iiG0bge6HkrqHPvBKjMeXGt__dCOoGG1Lw@mail.gmail.com/), but I think Michel Ludwig's answer only makes sense if you tried to relicense the cwl files under the MIT license. But since the MIT is (I think) GPL-compatible, you can just add the cwl files to your repo along with a notice that those files are licensed under the GPLv2, thus avoiding relicensing everything.

Of course you could relicense everything under the GPLv2 and save you some headaches, but I know I wouldn't want that for my project (c'mon, it's not even v3 !). Or follow Michel's advice and spam all the authors you can find and ask them for an exemption/relicense.

I plan to contact the main Kile authors and ask them if they would philosophically be okay with my solution : add the Kile cwl files/derived work to my repo, and release everything as a bundle, my code under the WTFPL (also GPL compatible) and the cwl files under the GPLv2. They are usually pretty sensible, and they didn't strike me as copyleft fanatics.

**Comment by btstream** [comment]
I have updated this pull request. Now, cwl file are separated into an independent package licensed under GPLv2.  I think this would be a better choice.

The package has not been published yet, and can be found at [https://github.com/btstream/latextools-cwl-completion-files](https://github.com/btstream/latextools-cwl-completion-files) . 

P.S. I'm really sorry for waiting so long to give response.

**Pull request closed without merging**